File: shared/mobile-sdk/src/commonMain/kotlin/io/sentient/mobilesdk/connectors/TaskListConnector.kt
// ---------------------------------------------------------------------------
// TaskListConnector — the composer task strip's mirror.
//
// Holds the last `tasklist.state` and nothing else. The gateway sends FULL
// STATE every time and owns row lifetime (a foreground call dies with its
// turn; a background delegateTask outlives it), so there is no upsert, no
// dedup and no clearing rule on this side. Replacing the list IS the contract:
// a replayed frame, a fan-out to a second window and a late joiner's attach are
// all the same operation.
//
// The one live-tool surface. Its retired predecessor mirrored a per-call
// tool-update frame: it kept tool rows forever (its `clear()` had no production
// caller) and left the UI deriving which bubble a pill belonged to — a question
// with no stable answer once a mid-turn steer splits a reply. Both that frame
// and that connector are gone; see shared/protocol/WIRE.md.
//
// Threading: single-threaded; the orchestrator routes frames on one dispatcher.
// ---------------------------------------------------------------------------
package io.sentient.mobilesdk.connectors

import io.sentient.mobilesdk.log.createLogger
import io.sentient.mobilesdk.protocol.ServerMessage
import io.sentient.mobilesdk.protocol.TaskListItem

class TaskListConnector(
    private val onUpdate: ((turnId: String?, items: List<TaskListItem>) -> Unit)? = null,
) : Connector {
    override val capability: String = CAPABILITY

    private val log = createLogger("connector", "task-list")

    private var items: List<TaskListItem> = emptyList()
    private var turn: String? = null

    fun list(): List<TaskListItem> = items

    fun turnId(): String? = turn

    override fun handle(msg: ServerMessage) {
        if (msg !is ServerMessage.TaskListState) return
        items = msg.items
        turn = msg.turnId
        // Row COUNT only — argsPreview is user content (PrivacyGuardTest).
        log.info("state", mapOf("turnId" to (msg.turnId ?: "-"), "count" to msg.items.size))
        onUpdate?.invoke(turn, items)
    }

    /**
     * Drop every row and re-fire [onUpdate] with the empty result. Driven by the
     * orchestrator's `clearConversationScopedState` from EVERY leave-the-conversation
     * entry point (mirrors [DelegationProgressConnector.clear]).
     *
     * The gateway's `tasklist.state` is a per-`SessionRuntime` projector that only
     * re-emits on its OWN mutations (turn start/tool-update/delegation-progress/turn
     * end) — `conversation.activate` triggers no resync push. Without this, switching
     * to a conversation with no task activity of its own leaves the client holding the
     * PREVIOUS conversation's last-known rows indefinitely: exactly the leak
     * [DelegationProgressConnector.clear]'s doc comment describes, and exactly why
     * this re-invokes [onUpdate] rather than silently resetting internal state the way
     * this connector's retired predecessor did (its `clear()` had no production caller
     * and left every consumer holding stale rows forever).
     */
    fun clear() {
        if (items.isEmpty() && turn == null) return
        log.info("clear", mapOf("count" to items.size))
        items = emptyList()
        turn = null
        onUpdate?.invoke(turn, items)
    }

    companion object {
        const val CAPABILITY: String = "tasklist"
    }
}

File: shared/web-sdk/src/connectors/task-list-connector.ts
import type { TaskListItem } from "@sentient/protocol";
import type { Connector, SentientSDKInternal } from "../connector-types.ts";
import { createLogger } from "../logger.ts";

const log = createLogger(["sentient", "sdk", "connectors", "task-list"]);

export interface TaskListConnectorConfig {
  /** Called on every `tasklist.state` frame, and on reset(). */
  onUpdate?: (turnId: string | null, items: readonly TaskListItem[]) => void;
}

// ---------------------------------------------------------------------------
// TaskListConnector — the composer task strip's mirror.
//
// Capability: "tasklist"
// Direction: status (observer; no outbound protocol from here)
//
// Holds whatever the last `tasklist.state` frame said and nothing else. The
// gateway sends FULL STATE every time (runtime/task-list.ts owns which rows
// exist and how long each lives), so there is no merge, no dedupe, and no
// lifetime rule on this side. Replacing the whole list is the entire contract:
// a replayed frame, a fan-out to a second window, and a late joiner's attach
// are the same operation.
//
// Replaces the retired ToolStatusConnector for new consumers (ToolStatusConnector
// itself stays in place until the webui migrates off it), which held tool
// calls forever and left every consumer deriving which bubble a pill belonged
// to.
// ---------------------------------------------------------------------------

export class TaskListConnector implements Connector {
  readonly capability = "tasklist";
  readonly kind = "status" as const;

  private readonly config: TaskListConnectorConfig;
  private unsubs: (() => void)[] = [];
  private items: readonly TaskListItem[] = [];
  private turn: string | null = null;

  constructor(config: TaskListConnectorConfig = {}) {
    this.config = config;
  }

  list(): readonly TaskListItem[] {
    return this.items;
  }

  turnId(): string | null {
    return this.turn;
  }

  attach(sdk: SentientSDKInternal): void {
    this.unsubs.push(
      sdk.onMessage("tasklist.state", (msg: unknown) => {
        const m = msg as { turnId?: string | null; items?: TaskListItem[] };
        this.items = Array.isArray(m.items) ? m.items : [];
        this.turn = m.turnId ?? null;
        // Row COUNT only — argsPreview is user content and never logged.
        log.debug("tasklist.state", { turnId: this.turn, count: this.items.length });
        this.config.onUpdate?.(this.turn, this.items);
      }),
    );
  }

  detach(): void {
    for (const unsub of this.unsubs) unsub();
    this.unsubs = [];
  }

  /** Drop everything. Called on disconnect, same as every other mirror. */
  reset(): void {
    const previousCount = this.items.length;
    this.items = [];
    this.turn = null;
    log.info("task-list.reset", { reason: "session identity teardown", previousCount });
    this.config.onUpdate?.(null, this.items);
  }
}
